### DISCUSSION TITLE
4C is available pre-built in a docker image

### DISCUSSION THREAD

**Discussion by ppraegla** [4C-multiphysics/4C#125] [Announcements]
With PR #96 we introduced a docker image with a pre-built 4C executable. Anyone can test and use 4C with just a few simple commands without configuring any environment and waiting for the build to complete. You can just pull the image `docker pull ghcr.io/4c-multiphysics/4c:latest`.

## Usage
Run the docker image
```
docker run --interactive --tty ghcr.io/4c-multiphysics/4c:latest
```
and, e.g., run the minimal tests
```
cd /home/user/4C/build
ctest -L minimal
```

### Run your input files
Run the docker image and mount a directory (e.g., the current directory `./`)
```
docker run --interactive --tty --mount type=bind,source=./,target=/workspace ghcr.io/4c-multiphysics/4c:latest
```
Set the correct owner of the mount (Unfortunately, we need to do that because we are the user user inside the docker image and the mount point is owned by root)
```
sudo chown user:user -R /workspace/
```
Run 4C with an input file from the host system and the output is also written to the host system
```
mpirun -np 2 /home/user/4C/build/4C /workspace/particle_dem_1d_radius_from_input.dat /workspace/out
```
Visualize the results in Paraview on your PC.

**Disclaimer** This whole process is still a little awkward to use. But there will certainly be improvements in the future.

## Additional information
At the moment, the docker image includes all dependencies required to run and test 4C, as well as the source. We will work on reducing the size of the docker image and only include the essentials to run 4C.

## Share your feedback
Please give feedback on how well the docker image works and what can be improved for your workflow.

**Comment by gilrrei** [comment]
This works quite nicely! What about introducing an alias for 4C within the container?

**Reply by gilrrei** [reply]
Or alternatively using entry points in docker :thinking:

**Reply by ppraegla** [reply]
Thanks for the suggestion. I will look into it at some point :)

**Comment by mayrmt** [comment]
I ran into problems with disk space, indicated by docker output such as

```
docker: failed to register layer: write /home/user/4C/build/src/red_airways/CMakeFiles/red_airways_objs.dir/4C_red_airways_implicitintegration.cpp.o: no space left on device

Run 'docker run --help' for more information
```

I could solve it by changing docker settings and increasing the allowed disk space for docker images.

**Comment by ischeider** [comment]
Maybe it's not a good idea to mount the personal project file to `/workspace` when running own input files; one should rather mount it to `/home/user/project`, i.e.
```
docker run --interactive --tty --mount type=bind,source=./,target=/home/user/project ghcr.io/4c-multiphysics/4c:latest
```
or, shorter:
```
docker run -it -v.:/home/user/project ghcr.io/4c-multiphysics/4c:latest
```
Then it is accessible without any permission issues. 

Actually, I don't know why we don't put this stuff to the documentation? In my opinion, all necessary details are included.

**Reply by ppraegla** [reply]
Is the other mount target working for you? When I run `docker run -it -v.:/home/user/project ghcr.io/4c-multiphysics/4c:latest`. I still can't create files inside the folder because it is owned by root. 

It is quite tricky to get the permissions right. I need to check if there is an easier solution.

**Reply by ischeider** [reply]
This is strange. I just ran a simulation, and all the result files have been written to my project directory.
```
user@18824f51c646:~$ ll -tr
total 76
-rw-r--r--  1 user user   807 Aug 30 11:50 .profile
-rw-r--r--  1 user user  3771 Aug 30 11:50 .bashrc
-rw-r--r--  1 user user   220 Aug 30 11:50 .bash_logout
drwxr-xr-x  1 root root  4096 Aug 30 11:50 ../
drwxr-xr-x  1 user user  4096 Sep  4 01:59 4C/
drwxr-xr-x  3 user user  4096 Sep  4 01:59 .cache/
drwxr-x---  1 user user  4096 Nov 14 10:44 ./
drwxrwxr-x 47 user user 40960 Nov 14 10:45 project/
```
Does the listing of the home directory look different for you?

**Reply by ppraegla** [reply]
Interesting. For me `project` is owned by root
```
-rw-r--r-- 1 user user  807 Aug 30 11:50 .profile
-rw-r--r-- 1 user user 3771 Aug 30 11:50 .bashrc
-rw-r--r-- 1 user user  220 Aug 30 11:50 .bash_logout
drwxr-xr-x 1 root root 4096 Aug 30 11:50 ../
drwxr-xr-x 1 user user 4096 Sep  4 01:59 4C/
drwxr-xr-x 3 user user 4096 Sep  4 01:59 .cache/
drwxr-x--- 1 user user 4096 Nov 14 10:35 ./
drwxrwxr-x 2 root root 4096 Nov 14 10:36 project/
```

**Reply by mayrmt** [reply]
I agree with @ischeider: docker stuff needs to be documented properly -- in the documentation, not in some closed discussion.

**Reply by ischeider** [reply]
@mayrmt Could you test, whether a mounted host directory is owned by root or user, when you run the container? I don't see the reason for a difference between @ppraegla and me.
